### PR TITLE
[FIX] Implement the --copyright flag, along with appropriate test cases.

### LINKED ISSUE #687: Flag --copyright does not exist, even though it's referenced in the code.
Resolve by adding a --copyright flag which prints the right copyright/license info.

### PR DESCRIPTION
**Pull request by joshuak94**
Resolves #687 by creating a format_copyright and the appropriate tests.

### CODE REVIEW

File: include/seqan3/argument_parser/detail/format_help.hpp [hunk 1]
```diff
@@ -764,4 +765,84 @@ class format_version : public format_help
...
+            std::cout << to_text("\\fP") << meta.short_copyright << "\n\n";
+        }
+        else
+            std::cout << to_text("\\fP") << meta.app_name << "copyright information not available.\n\n";
+
+        std::cout << to_text("\\fB") << "SeqAn3 Copyright:\n";
+        std::string line;
+        while (std::getline(seqan_license, line))
+        {
+            std::fill_n(out, layout.leftPadding, ' ');
```
Comment (smehringer) [inline include/seqan3/argument_parser/detail/format_help.hpp]: I am pretty sure that for this line and the one below you can just use the print_line function of the format
```cpp
print_line(line);
```
Maybe you need `print_line(line, false);`.

Comment (joshuak94) [reply include/seqan3/argument_parser/detail/format_help.hpp]: I tried it, but with `print_line(line)` it treats every line as a paragraph and makes the license info huge. But with `print_line(line, false)` it treats the whole thing as one paragraph and removes the spacing in between... So maybe it's better to just keep it as is?

File: include/seqan3/argument_parser/detail/format_help.hpp [hunk 2]
```diff
@@ -764,4 +765,84 @@ class format_version : public format_help
...
+ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+DAMAGE.)"};
+
+        print_header();
```
[the remark this answers is not in the document]
Comment (joshuak94) [reply include/seqan3/argument_parser/detail/format_help.hpp]: Would it be better to give a width of `layout.screenWidth`?

**Comment by joshuak94** [comment]
@smehringer Implemented the license file.

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 3]
```diff
@@ -105,16 +105,67 @@ TEST(help_add_test, add_option)
...
-    // const char * argvCopyright[] = {"./copyright", "--copyright"};
-    // argument_parser copyright("copyright", 2, argvCopyright);
-    // copyright.info.long_copyright = "long copyright";
-    // testing::internal::CaptureStdout();
-    // EXPECT_THROW(copyright.parse(), parser_interruption);
-    // stdout = testing::internal::GetCapturedStdout();
-    // expected = "whatever is expected";
-    // EXPECT_TRUE(ranges::equal((stdout   | ranges::view::remove_if(is_space)),
-                               // expected | ranges::view::remove_if(is_space)));
+    // Tests the --copyright call.
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]: Please put this into its own test case instead of adding it to the `TEST(help_add_test, add_option)` test.

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 4]
```diff
@@ -105,16 +105,67 @@ TEST(help_add_test, add_option)
...
+                               "--------------------------------------------------------------------------------"
+                               "copyright copyright information not available."
+                               "================================================================================"
+                               "This program contains SeqAn3 code licensed under the following terms:"
+                               "--------------------------------------------------------------------------------"
+                               + license_string);
+        EXPECT_TRUE(ranges::equal((stdout   | ranges::view::remove_if(is_space)),
+                                   expected | ranges::view::remove_if(is_space)));
+    }
+    copyright.info.short_copyright = "short copyright\nshortcopyright";
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]: swap this with line below so the explanation comes first and introduce a newline between the tests. :nail_care:  makes reading easier

**Comment by joshuak94** [comment]
@smehringer Updated!

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 5]
```diff
@@ -105,31 +107,20 @@ TEST(help_add_test, add_option)
...
-
     // Empty help call with -hh
     const char * argv2[] = {"./help_add_test", "-hh"};
     argument_parser parser2("test_parser_2", 2, argv2);
     testing::internal::CaptureStdout();
     EXPECT_THROW(parser2.parse(), parser_interruption);
     stdout = testing::internal::GetCapturedStdout();
     expected = std::string("test_parser_2"
-                           "============="
-                           "VERSION"
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]: remove this change again :nail_care:

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 6]
```diff
@@ -138,13 +129,13 @@ TEST(help_add_test, add_option)
     EXPECT_THROW(parser3.parse(), parser_interruption);
     stdout = testing::internal::GetCapturedStdout();
     expected = std::string("version"
-                           "======="
-                           "VERSION"
-                           "Last update:"
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]: here too

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 7]
```diff
@@ -156,15 +147,15 @@ TEST(help_add_test, add_option)
     EXPECT_THROW(parser4.parse(), parser_interruption);
     stdout = testing::internal::GetCapturedStdout();
     expected = std::string("versionURL"
-                           "=========="
-                           "VERSION"
-                           "Last update:"
-                           "versionURL version:"
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]: here too :nerd_face:

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 8]
```diff
@@ -174,14 +165,14 @@ TEST(help_add_test, add_option)
     EXPECT_THROW(parser5.parse(), parser_interruption);
     stdout = testing::internal::GetCapturedStdout();
     expected = std::string("hidden"
-                           "======"
-                           "OPTIONS"
-                           "VERSION"
-                           "Last update:"
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]: aaaaand here

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 9]
```diff
@@ -199,30 +190,96 @@ TEST(help_add_test, add_option)
...
-                           "full synopsis"
-                           "full synopsis2"
-                           "DESCRIPTION"
-                           "description"
-                           "description2"
-                           "POSITIONAL ARGUMENTS"
-                           "ARGUMENT-1 List of STRING's"
-                           "this is a positional option."
-                           "OPTIONS"
-                           "-i, --int INT (32 bit)"
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]: and since i commented everywhere.. here too :grin:

**Comment by joshuak94** [comment]
@smehringer **now** everything should be okay!

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 10]
```diff
@@ -223,6 +214,72 @@ TEST(help_add_test, add_option)
...
+                                std::istreambuf_iterator<char>()};
+    // Test --copyright with empty short and long copyright info.
+    {
+        testing::internal::CaptureStdout();
+        EXPECT_THROW(copyright.parse(), parser_interruption);
+        stdout = testing::internal::GetCapturedStdout();
+        expected = std::string("================================================================================"
+                               "Copyright information for copyright:"
+                               "--------------------------------------------------------------------------------"
+                               "copyright copyright information not available."
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]: there is a double copyright? If this is the app name please change the app name to something like `MyApp` so it is clearer and also I would propose a `:`
-> `MayApp: copyright information not available.`

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 11]
```diff
@@ -223,6 +214,72 @@ TEST(help_add_test, add_option)
...
+    {
+        testing::internal::CaptureStdout();
+        EXPECT_THROW(copyright.parse(), parser_interruption);
+        stdout = testing::internal::GetCapturedStdout();
+        expected = std::string("================================================================================"
+                               "Copyright information for copyright:"
+                               "--------------------------------------------------------------------------------"
+                               "copyright full copyright information not available. Displaying short copyright information instead:"
+                               "short copyright"
+                               "short copyright"
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]: double short copy right ? ah I see that this is intentional. can you change it to 
`short copyright line 1\n short copy right line 2`? otherwise people (like me) might think this is an error.

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 12]
```diff
@@ -223,6 +214,72 @@ TEST(help_add_test, add_option)
...
+    // Test --copyright with a non-empty short copyright and a non-empty long copyright.
+    copyright.info.long_copyright = "long copyright\nlong copyright";
+    {
+        testing::internal::CaptureStdout();
+        EXPECT_THROW(copyright.parse(), parser_interruption);
+        stdout = testing::internal::GetCapturedStdout();
+        expected = std::string("================================================================================"
+                               "Copyright information for copyright:"
+                               "--------------------------------------------------------------------------------"
+                               "copyright Copyright:"
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]: actually why do you have `copyright Copyright:` here if in the line above you already have `Copyright information for copyright:"`? isn't that redundant?

Comment (joshuak94) [reply test/unit/argument_parser/detail/format_help_test.cpp]: Good point. Removed it!

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 13]
```diff
@@ -223,6 +214,71 @@ TEST(help_add_test, add_option)
...
+}
+
+TEST(test_add_copyright, copyright)
+{
+    std::string stdout;
+    std::string expected;
+
+    // Tests the --copyright call.
+    const char * argvCopyright[] = {"./copyright", "--copyright"};
+    argument_parser copyright("copyright", 2, argvCopyright);
```
Comment (smehringer) [inline test/unit/argument_parser/detail/format_help_test.cpp]:
```suggestion
    argument_parser copyright("myApp", 2, argvCopyright);
```
thats probably why your test fails

Comment (joshuak94) [reply test/unit/argument_parser/detail/format_help_test.cpp]: Whoops, was a weird thing that happened while rebasing. Thanks!

**Comment by smehringer** [comment]
@h-2 review ping

File: test/unit/argument_parser/detail/CMakeLists.txt [hunk 14]
```diff
@@ -1,2 +1,3 @@
+add_definitions(-DSEQAN_INCLUDE_DIR="${SEQAN3_INCLUDE_DIR}")
```
Comment (h-2) [inline test/unit/argument_parser/detail/CMakeLists.txt]: why was this change necessary?

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 15]
```diff
@@ -45,11 +47,11 @@ TEST(help_add_test, add_option)
...
-                                       "==========="
-                                       "VERSION"
-                                       "Last update:"
-                                       "test_parser version:"
-                                       "SeqAn version: 3.0.0");
+                           "==========="
+                           "VERSION"
+                           "Last update:"
+                           "test_parser version:"
+                           "SeqAn version: 3.0.0");
```
Comment (h-2) [inline test/unit/argument_parser/detail/format_help_test.cpp]: include the actual seqan3 version from seqan3/version.hpp instead of hard-coding

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 16]
```diff
@@ -223,6 +214,71 @@ TEST(help_add_test, add_option)
                            "SeqAn version: 3.0.0");
     EXPECT_TRUE(ranges::equal((stdout   | ranges::view::remove_if(is_space)),
                                expected | ranges::view::remove_if(is_space)));
```
Comment (h-2) [inline test/unit/argument_parser/detail/format_help_test.cpp]: use view::filter(!...) instead view::remove_if(...) because the latter is deprecated.

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 17]
```diff
@@ -223,6 +214,71 @@ TEST(help_add_test, add_option)
...
-   // EXPECT_THROW(parser6.parse(), parser_interruption);
+    std::ifstream license_file{std::string{{SEQAN_INCLUDE_DIR}} + "/../LICENSE"};
+    std::string license_string{(std::istreambuf_iterator<char>(license_file)),
+                                std::istreambuf_iterator<char>()};
+    // Test --copyright with empty short and long copyright info.
+    {
+        testing::internal::CaptureStdout();
+        EXPECT_THROW(copyright.parse(), parser_interruption);
+        stdout = testing::internal::GetCapturedStdout();
+        expected = std::string("================================================================================"
```
Comment (h-2) [inline test/unit/argument_parser/detail/format_help_test.cpp]: why is this wrapped in `std::string()`?

Comment (h-2) [reply test/unit/argument_parser/detail/format_help_test.cpp]: also in other places

File: test/unit/argument_parser/detail/CMakeLists.txt (hunk 14 above)
Comment (smehringer) [reply test/unit/argument_parser/detail/CMakeLists.txt]: see the commit message. Because we want to read the licence file (and therefore we need the repo path) in order to compare it to the hard-coded copyright printing and keep things updated.
